autocertkit/utils.py

```diff
@@ -1183,29 +1184,45 @@ def ping(vm_ip, dst_vm_ip, interface, packet_size=1400,
                         % result)
 
 
+@log_exceptions
 def ssh_command(ip, username, password, cmd_str, dbg_str=None, attempts=10):
-    """execute an SSH command using the parimiko library, in order
-    to specify a password. Return the result to the caller."""
+    """execute an SSH command using the parimiko library, return both
+    exit code, stdout and stderr."""
     if dbg_str:
         log.debug(dbg_str)
 
+    # use uuid as random string
+    flag = "__%s__" % str(uuid.uuid4())
+    of = "/tmp/ack.stdout.%s.txt" % flag
+    ef = "/tmp/ack.stderr.%s.txt" % flag
+
+    # put cmd_str into single quotes and sh to run with redirection, e.g.
+    #   echo "$HOME" | sed 's/\///g'
+    #   sh -c 'echo "$HOME" | sed '\''s/\///g'\''' 1>of 2>ef
+    cmd_s = cmd_str.replace("'", r"'\''")
+    cmd_s = '''sh -c '%s' 1>%s 2>%s''' % (cmd_s, of, ef)
+    # catch exit code, stdout and stderr
+    cmd = '''%s; echo -n "$?%s"; cat %s; echo -n "%s"; cat %s; rm -f %s %s;''' % \
```

Comment:
I don't quite get why we are wanting to force stdout/stderr into a single string to then split it out again? why not keep the output separately as returned from the SSH command?

Reply:
Please see make_local_call() in autocertkit
```
def make_local_call(call, logging=True, std_out=subprocess.PIPE, shell=False):
    """Function wrapper for making a simple call to shell"""
    if logging:
        log.debug("make_local_call: %s" % (" ".join(call)))
    process = subprocess.Popen(call, stdout=std_out, shell=shell)
    stdout, stderr = process.communicate()
    res = {"returncode": process.returncode, "stdout": str(
        stdout).strip(), "stderr": str(stderr).strip()}
    if logging:
        log.debug("returncode: %d" % process.returncode)
        log.debug("stdout: %s" % str(stdout))
        log.debug("stderr: %s" % str(stderr))
        if res["returncode"] != 0:
            msg = "ERR: Could not make local call %s" % ' '.join(call)
            log.debug(msg)

    return res
```
I like the style, it's a fundamental function to run command with exit code, stdout and stderr separated return. 
In original implement ssh_command(), it's inappropriate as low level infrastructure
```
def ssh_command(ip, username, password, cmd_str, dbg_str=None, attempts=10):
```

I know at present almost all ACK existing code do not check command exit code, and just capture  stdout + stderr, which is not rigorous logic.
I refactor ssh_command() is to 
- make it should do
- add and unify the run-command plugin, no matter run on host locally or on guest remotely
- add the ability of run any complex shell on guest VM, and no need to add new plugin interface.
- make the prepare for collect guest 'bug-report' to help locate ACK test issues. You know at present ACK would not collect guest info (config, logs, etc.)

Reply:
OK I see that the issue here is that we were _not_ returning stderr/return code etc.

Would it not be better to modify the `sshCommand` call to return these values? That way you save having to have this overcomplicated logic combining then separating out these various values?

E.g. here:
https://github.com/xenserver/auto-cert-kit/blob/develop/autocertkit/ssh.py#L189

Paramiko supports returning them separately, I think that just needs to be plumbed through.

Reply:
Understood. I noticed this before, and I gave up because the module ssh.py is complex enough, that needs sufficient time to investigate also paramiko. I think I can refactor it later, for it's low priority at present.